api/services/update/state.js

```diff
@@ -0,0 +1,148 @@
+  try {
+    state = JSON.parse(state);
+  } catch (err) {
+    logger.error(`getState on JSON.parse: ${err}`);
+  }
+  return state;
+};
+
+/**
+ * write the latest version of the state to the file
+ * @param {object} state - state in JSON format
+ * @param {object} filename - name of the file where the state is saved
+ */
+const updateStateInFile = async (state, filename) => {
+  const pathfile = path.resolve(stateDir, filename);
+  const isPathExist = await fs.pathExists(pathfile);
+  if (!isPathExist) {
+    logger.error(`updateStateInFile on fs.pathExists: file ${pathfile} doesn't exist`);
+  } else {
+    await fs.writeFile(pathfile, JSON.stringify(state, null, 2));
+  }
+};
+
+/**
+ * add step "askUnpaywall" in steps attributes of state
+ * @param {string} filename - name of the file where the state is saved
+ */
+const addStepAskUnpaywall = async (filename) => {
+  const state = await getState(filename);
+  logger.info('step - ask unpaywall');
+  const step = {
+    task: 'askUnpaywall',
+    took: 0,
+    status: 'inProgress',
+  };
+  state.steps.push(step);
+  await updateStateInFile(state, filename);
+};
+
+/**
+ * add step "download" in steps attributes of state
+ * @param {string} filename - name of the file where the state is saved
+ * @param {string} downloadFile - unpaywall data update file name
+ */
+const addStepDownload = async (filename, downloadFile) => {
+  const state = await getState(filename);
+  logger.info('step - download file');
+  const step = {
+    task: 'download',
+    file: downloadFile,
+    percent: 0,
+    took: 0,
+    status: 'inProgress',
+  };
+  state.steps.push(step);
+  await updateStateInFile(state, filename);
+};
+
+/**
+ * add step "download" in steps attributes of state
+ * @param {string} filename - name of the file where the state is saved
+ * @param {string} downloadFile - unpaywall data update file name
+ */
+const addStepInsert = async (filename, downloadFile) => {
+  const state = await getState(filename);
+  logger.info('step - insert file');
+  const step = {
+    task: 'insert',
+    file: downloadFile,
+    linesRead: 0,
+    percent: 0,
+    took: 0,
+    status: 'inProgress',
+  };
+  state.steps.push(step);
+  await updateStateInFile(state, filename);
+};
+
+/**
+ * update the state when there is an error
+ * @param {string} filename - name of the file where the state is saved
+ */
+const fail = async (filename) => {
+  const state = await getState(filename);
+  state.done = true;
+  state.endAt = new Date();
+  state.error = true;
+  await updateStateInFile(state, filename);
+};
+
+/**
+ * update the state when the process is finished
+ * @param {string} filename - name of the file where the state is saved
+ */
+const endState = async (filename) => {
+  const state = await getState(filename);
+  state.done = true;
+  state.endAt = new Date();
+  state.took = (new Date(state.endAt) - new Date(state.createdAt)) / 1000;
```

Comment:
`endAt` and `createdAt` are already `Date` objects, so I guess there's no need to cast them into Date.